system/jlib/jsocket.cpp

```diff
@@ -1921,257 +1921,149 @@ int CSocket::wait_write(unsigned timeout)
+
+        if (rc > 0)
+        {
+            sizeRead += rc;
+            if (sizeRead >= min_size)
+                break;
         }
-        if (rc < 0) {
-            int err = SOCKETERRNO();
-            if (BADSOCKERR(err)) {
-                // don't think this should happen but convert to same as shutdown while investigation
-                LOGERR2(err,1,"Socket closed during read");
-                rc = 0;
-            }
-            else if ((err==JSE_INTR)&&(retrycount--!=0)) {
-                if (sock==INVALID_SOCKET)
-                    rc = 0;         // convert an EINTR after closed to a graceful close
+        else
+        {
+            if (rc < 0)
+            {
+                int err = SOCKETERRNO();
+                if (BADSOCKERR(err))
+                {
+                    // don't think this should happen but convert to same as shutdown while investigation
+                    LOGERR2(err,1,"Socket closed during read");
+                    rc = 0;
+                }
+                else if ((err==JSE_INTR)&&(retrycount--!=0))
+                {
+                    if (sock==INVALID_SOCKET)
+                        rc = 0;         // convert an EINTR after closed to a graceful close
+                    else
+                    {
+                        LOGERR2(err,1,"EINTR retrying");
+                        goto EintrRetry;
+                    }
+                }
                 else
                 {
-                    LOGERR2(err,1,"EINTR retrying");
-                    goto EintrRetry;
+                    if (nonblocking && (err == JSE_WOULDBLOCK || err == EAGAIN)) // if EGAIN or EWOULDBLOCK - no more data to read
+                    {
+                        if (0 == min_size) // if min_read is 0, then whatever we have read so far is good enough (even if 0)
+                            break;
+                        // fall through/loop around. NB: rc != 0
+                    }
+                    else
+                    {
+                        VStringBuffer errMsg("readtms(timeoutMs=%d)", timeoutMs);
+                        LOGERR2(err,1,errMsg.str());
+                        if ((err==JSE_CONNRESET)||(err==JSE_INTR)||(err==JSE_CONNABORTED))
+                        {
+                            errclose();
+                            err = JSOCKERR_broken_pipe;
+                        }
+                        THROWJSOCKTARGETEXCEPTION(err);
+                    }
                 }
             }
-            else {
-                VStringBuffer errMsg("readtms(timeoutms=%d)", timeoutms);
-                LOGERR2(err,1,errMsg.str());
-                if ((err==JSE_CONNRESET)||(err==JSE_INTR)||(err==JSE_CONNABORTED)) {
-                    errclose();
-                    err = JSOCKERR_broken_pipe;
-                }
-                THROWJSOCKEXCEPTION(err);
+            if (rc == 0)
+            {
+                state = ss_shutdown;
+                if (min_size==0) 
```

Comment:
I think in this case, yes, that may be better/more correct.
This condition is suppressing the graceful close in the specific case when min_size is 0 at the moment.

This kept the previous semantics.
I think it would be safe to change and do so if sizeRead >= min_size here and in securesocket 

I'll change it.